monix-reactive/shared/src/test/scala/monix/reactive/internal/operators/MergePrioritizedListSuite.scala

```diff
@@ -0,0 +1,171 @@
+    }
+    assert(s.state.tasks.isEmpty, "tasks.isEmpty")
+  }
+
+  test("should push all items downstream before calling onComplete") { implicit s =>
+    val source =
+      Observable.mergePrioritizedList((1, Observable.now(1L)), (1, Observable.now(1L)), (1, Observable.now(1L)))
+    var count = 0L
+    source.unsafeSubscribeFn(new Observer[Long] {
+      def onNext(elem: Long): Future[Ack] = {
+        Task
+          .sleep(FiniteDuration(1, SECONDS))
+          .map { _ =>
+            count += elem
+            Continue
+          }
+          .runToFuture
+      }
+      def onError(ex: Throwable): Unit = throw ex
+      def onComplete(): Unit = assertEquals(count, 3L)
+    })
+
+    s.tick(1.seconds)
+    assertEquals(count, 1L)
+    s.tick(1.seconds)
+    assertEquals(count, 2L)
+    s.tick(1.seconds)
+    assertEquals(count, 3L)
+    assert(s.state.tasks.isEmpty, "tasks.isEmpty")
+  }
+
+  test("should complete all upstream onNext promises when downstream stops early") { implicit s =>
+    val sources = (1 to 10).map(i => (i, new OnNextExposingObservable(i * 1L)))
+    val source = Observable.mergePrioritizedList(sources: _*)
+
+    source.unsafeSubscribeFn(new Observer[Long] {
+      def onNext(elem: Long): Future[Ack] = {
+        Task
+          .sleep(FiniteDuration(1, SECONDS)) // sleep so that other source observables' items queued before next call
+          .map { _ =>
+            Stop
+          }
+          .runToFuture
+      }
+      def onError(ex: Throwable): Unit = throw ex
+      def onComplete(): Unit = ()
+    })
+
+    s.tick(1.seconds)
+    sources.foreach(src => assert(src._2.onNextRes.exists(_.isCompleted), "source promise completed"))
+  }
+
+  test("should complete all upstream onNext promises when downstream errors early") { implicit s =>
+    val sources = (1 to 10).map(i => (i, new OnNextExposingObservable(i * 1L)))
+    val source = Observable.mergePrioritizedList(sources: _*)
+
+    source.unsafeSubscribeFn(new Observer[Long] {
+      def onNext(elem: Long): Future[Ack] = {
+        Task
+          .sleep(FiniteDuration(1, SECONDS)) // sleep so that other source observables' items queued before next call
+          .map { _ =>
+            throw new RuntimeException("downstream failed")
+          }
+          .runToFuture
+      }
+      def onError(ex: Throwable): Unit = throw ex
+      def onComplete(): Unit = ()
+    })
+
+    s.tick(1.seconds)
+    sources.foreach(src => assert(src._2.onNextRes.exists(_.isCompleted), "source promise completed"))
+  }
+
+  // Enables verification that MergePrioritizedListObservable completes
+  // upstream onNext promises when downstream stops early
+  private class OnNextExposingObservable[+A](elem: A)(implicit s: Scheduler) extends Observable[A] {
+    var onNextRes: Option[Future[Ack]] = None
+
+    def unsafeSubscribeFn(subscriber: Subscriber[A]): Cancelable = {
+      val onNext = subscriber.onNext(elem)
+      onNext.onComplete(_ => subscriber.onComplete())
```

Comment:
since it's allowed in the contract, it might be a better test to change this to call `onComplete` immediately, rather than waiting for `onNext` to complete. Having tried it, the test still passes, so it's really just a case of making it a little bit more stringent

Reply:
Done